lisa/tools/kdump.py

```diff
@@ -0,0 +1,326 @@
+
+    @property
+    def can_install(self) -> bool:
+        return True
+
+    def _install(self) -> bool:
+        tool_name = KernelDumpTool(self.node.os).get_tool_package_name("makedumpfile")
+        posix_os: Posix = cast(Posix, self.node.os)
+        posix_os.install_packages(tool_name)
+        return self._check_exists()
+
+
+class Kdump(Tool):
+    """
+    kdump is a feature of the Linux kernel that creates crash dumps in the event of a
+    kernel crash. When triggered, kdump exports a memory image (also known as vmcore)
+    that can be analyzed for the purposes of debugging and determining the cause of a
+    crash.
+
+    kdump tool manages the kdump feature of the Linux kernel. Different distro os has
+    different kdump tool.
+
+    We can support Redhat, Suse, Debian family distro now. We will add support for
+    other distros.
+    """
+
+    # If the file /sys/kernel/kexec_crash_loaded does not exist. This means that the
+    # currently running kernel either was not configured to support kdump, or that a
+    # crashkernel= commandline parameter was not used when the currently running kernel
+    # booted. Value "1" means crash kernel is loaded, otherwise not loaded.
+    #
+    # It also has /sys/kernel/kexec_crash_size file, which record the crash kernel size
+    # of memory reserved. We don't need to check this file in our test case.
+    kexec_crash = "/sys/kernel/kexec_crash_loaded"
+
+    # This file shows you the current map of the system's memory for each physical
+    # device. We can check /proc/iomem file for memory reserved for crash kernel.
+    iomem = "/proc/iomem"
+
+    # Following are the configuration setting required for system and dump-capture
+    # kernels for enabling kdump support.
+    required_kernel_config = [
+        "CONFIG_KEXEC",
+        "CONFIG_CRASH_DUMP",
+        "CONFIG_PROC_VMCORE",
+    ]
+
+    dump_path = "/var/crash"
+    crashkernel_cfg_file = ""
+
+    @property
+    def dependencies(self) -> List[Type[Tool]]:
+        return [Kexec, Makedumpfile]
+
+    @property
+    def command(self) -> str:
+        return KernelDumpTool(self.node.os).get_tool_cmd_name("kdump")
+
+    @property
+    def can_install(self) -> bool:
+        return True
+
+    def _install(self) -> bool:
+        tool_name = KernelDumpTool(self.node.os).get_tool_package_name("kdump")
+        posix_os: Posix = cast(Posix, self.node.os)
+        posix_os.install_packages(tool_name)
+        self.enable_kdump_service()
+        return self._check_exists()
+
+    def check_required_kernel_config(self, config_path: str) -> None:
+        for config in self.required_kernel_config:
+            result = self.node.execute(f"grep {config}=y {config_path}")
+            result.assert_exit_code(
+                message=f"The kernel config {config} is not set."
+                "Kdump is not supported."
+            )
+
+    def _get_crashkernel_cfg_file(self) -> None:
+        if isinstance(self.node.os, Redhat):
```

Comment:
The logic is complex unnecessary, not easy to read, try below,

```python
if isinstance(self.node.os, Debian):
    self.crashkernel_cfg_file = "/etc/default/grub.d/kdump-tools.cfg"
elif isinstance(self.node.os, Redhat) and self.node.os.information.version >= "8.0.0"
    self.crashkernel_cfg_file = ""
else:
    self.crashkernel_cfg_file = "/etc/default/grub"

if self.crashkernel_cfg_file:
    assert self.node.shell.exists(PurePosixPath(self.crashkernel_cfg_file)), (
                    f"{self.crashkernel_cfg_file} doesn't exist. "
                    "Please check the right grub file for "
                    f"{self.node.os.name} {self.node.os.information.version}."
             )

```

Reply:
Done